File: 541. Reverse String II/main.cpp
#include <algorithm>
#include <string>

using std::reverse;
using std::string;

class Solution {
public:
    string reverseStr(string s, int k) {
        int len = s.size();
        for (int i = 0; i < len; i += 2 * k) {
            int min = k > len - i ? len - i : k;
            for (int j = 0; j < (min + 1) / 2; j++) {
                char temp = s[i + j];
                s[i + j] = s[i + min - 1 - j];
                s[i + min - 1 - j] = temp;
            }
        }

        return s;
    }
};

class Solution {
public:
    string reverseStr(string s, int k) {
        int len = s.size();
        for (int i = 0; i < len; i += 2 * k) {
            if (i + k < len) {
                reverse(s.begin() + i, s.begin() + i + k);
            } else {
                reverse(s.begin() + i, s.end());
            }
        }

        return s;
    }
};

File: 541. Reverse String II/main.rs
impl Solution {
    pub fn reverse_str(s: String, k: i32) -> String {
        s.chars()
            .collect::<Vec<char>>()
            .chunks_mut(k as usize)
            .enumerate()
            .flat_map(|(i, chunk)| {
                if i % 2 == 0 {
                    chunk.reverse();
                }
                chunk.iter().cloned()
            })
            .collect()
    }
}

impl Solution {
    pub fn reverse_str(s: String, k: i32) -> String {
        let mut vec = s.chars().collect::<Vec<char>>();

        for i in (0..vec.len()).step_by(2 * k as usize) {
            let end = std::cmp::min(i + k as usize, vec.len());
            vec[i..end].reverse();
        }

        vec.iter().collect()
    }
}

impl Solution {
    pub fn reverse_str(s: String, k: i32) -> String {
        let mut vec = s.chars().collect::<Vec<char>>();

        for i in (0..vec.len()).step_by(2 * k as usize) {
            let end = std::cmp::min(i + k as usize, vec.len());
            let mid = ((end - i + 1) / 2) as usize;
            for j in 0..mid {
                let c = vec[i + j];
                vec[i + j] = vec[end - 1 - j];
                vec[end - 1 - j] = c;
            }
        }

        vec.iter().collect()
    }
}
